[DNM] All tasks without dependencies are root-ish

gjoseph92 opened the pull request:
If a task doesn't have dependencies, then it's obviously a root task. However, the current logic would only consider it so if the TaskGroup was also 2x larger than the cluster.

That allowed for the awkward case where small groups of root tasks would go down a different code path. It's not clear how much this code path was even used or needed, xref https://github.com/dask/distributed/pull/6974.

Opening this just to see what happens in CI. 

See a little other discussion in https://github.com/dask/distributed/issues/7204#issuecomment-1293854506.

Making this change would help a bit with https://github.com/dask/distributed/issues/7197.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

mrocklin commented:
>That allowed for the awkward case where small groups of root tasks would go down a different code path. It's not clear how much this code path was even used or needed, xref https://github.com/dask/distributed/pull/6974.

I don't have a ton to add to this conversation, but some examples of important cases that look like this include:

1.  `da.from_zarr` where there is a single zarr array dataset in the graph
2. client.submit (possibly in a for loop)

Would it make sense to avoid this logic in these cases?

gjoseph92 commented:
First, let's clarify that this `is_rootish` check was originally written in https://github.com/dask/distributed/pull/4967 just to decide which tasks should get co-assigned, versus which should follow "normal" scheduling logic. (Then, within the "normal" scheduling logic, there was _another_ branch for the no-deps case we're talking about removing in https://github.com/dask/distributed/pull/6974.)

We've now co-opted this "should this task be co-assigned" logic and are also using it to decide "should this task be queued?" We do this because "is this a root task" happens to be a good answer to both of those questions. But when thinking about possible impacts, it could be helpful to ask "what's the effect of co-assigning these tasks" separately from "what's the effect of queuing these tasks".

1. `from_zarr`
    * co-assigning: doesn't matter. There's only one task anyway.
    * queuing: makes behavior more consistent. If there are other tasks already on the cluster, it makes more sense that this should wait in the queue until space opens up (though it has no effect on actual performance, since opening a zarr dataset doesn't take much memory).
1. single `client.submit`
    * co-assigning: doesn't matter. There's only one task. However, you lose round-robin. (Though `test_quiet_cluster_round_robin` still passes on this branch even though we're bypassing the round-robin code path, which I can't explain).
    * queuing: makes behavior more consistent. These tasks now have to wait in the queue like everything else instead of skipping in line.
1. `client.submit` in a for loop
    * co-assign: well, co-assignment has always been strange for this case. Before, we'd submit the first nthreads*2 tasks randomly/round-robin, then flip to co-assignment mode and submit them in batches to a worker. Now, we'll always submit in batches, but initially the batch size will be 1 until the TaskGroup grows large enough. So worker selection/load distribution will be very slightly different, but I think not in a way that matters or is noticeable.
    * queuing: makes behavior more consistent. All tasks are queued instead of the first nthreads*2 tasks getting a fast-pass.
1. `client.submit` in a for loop with follow-up tasks
    * co-assign: same as above, no meaningful effect
    * queuing: could reduce overproduction (assuming tasks are slower than task submission). Right now you'll get root task overproduction with submit in a for-loop, because the first nthreads*2 tasks will be sent to workers immediately before queuing flips on. If you `client.submit` all your root tasks, then `client.submit` downstream tasks before any root tasks have finished, overproduction will be prevented with this change.

**tl;dr I can't think of a meaningful way those cases would get worse, or reason they should avoid queuing/co-assignment logic**

Overall, I don't think this change should affect actual use much. Nearly all of my motivation for doing it is to reduce the number of branches to worry about, make things more consistent, simplify testing, and increase confidence that our tests are actually testing the queuing code path. For instance, because most tests just `client.submit` a couple tasks, a ton of tests were actually not using the queuing code path even with `worker-saturation: 1.0`. (This change is how I found https://github.com/dask/distributed/issues/7223).

mrocklin commented:
I'm not a blocker on this

On Mon, Oct 31, 2022 at 2:59 PM Gabe Joseph ***@***.***>
wrote:

> First, let's clarify that this is_rootish check was originally written in
> #4967 <https://github.com/dask/distributed/pull/4967> just to decide
> which tasks should get co-assigned, versus which should follow "normal"
> scheduling logic. (Then, within the "normal" scheduling logic, there was
> *another* branch for the no-deps case we're talking about removing in
> #6974 <https://github.com/dask/distributed/pull/6974>.)
>
> We've now co-opted this "should this task be co-assigned" logic and are
> also using it to decide "should this task be queued?" We do this because
> "is this a root task" happens to be a good answer to both of those
> questions. But when thinking about possible impacts, it could be helpful to
> ask "what's the effect of co-assigning these tasks" separately from "what's
> the effect of queuing these tasks".
>
>    1. from_zarr
>       - co-assigning: doesn't matter. There's only one task anyway.
>       - queuing: makes behavior more consistent. If there are other tasks
>       already on the cluster, it makes more sense that this should wait in the
>       queue until space opens up (though it has no effect on actual performance,
>       since opening a zarr dataset doesn't take much memory).
>    2. single client.submit
>       - co-assigning: doesn't matter. There's only one task. However, you
>       lose round-robin. (Though test_quiet_cluster_round_robin still
>       passes on this branch even though we're bypassing the round-robin code
>       path, which I can't explain).
>       - queuing: makes behavior more consistent. These tasks now have to
>       wait in the queue like everything else instead of skipping in line.
>    3. client.submit in a for loop
>       - co-assign: well, co-assignment has always been strange for this
>       case. Before, we'd submit the first nthreads*2 tasks randomly/round-robin,
>       then flip to co-assignment mode and submit them in batches to a worker.
>       Now, we'll always submit in batches, but initially the batch size will be 1
>       until the TaskGroup grows large enough. So worker selection/load
>       distribution will be very slightly different, but I think not in a way that
>       matters or is noticeable.
>       - queuing: makes behavior more consistent. All tasks are queued
>       instead of the first nthreads*2 tasks getting a fast-pass.
>    4. client.submit in a for loop with follow-up tasks
>       - co-assign: same as above, no meaningful effect
>       - queuing: could reduce overproduction (assuming tasks are slower
>       than task submission). Right now you'll get root task overproduction with
>       submit in a for-loop, because the first nthreads*2 tasks will be sent to
>       workers immediately before queuing flips on. If you client.submit
>       all your root tasks, then client.submit downstream tasks before any
>       root tasks have finished, overproduction will be prevented with this change.
>
> *tl;dr I can't think of a meaningful way those cases would get worse, or
> reason they should avoid queuing/co-assignment logic*
>
> Overall, I don't think this change should affect actual use much. Nearly
> all of my motivation for doing it is to reduce the number of branches to
> worry about, make things more consistent, simplify testing, and increase
> confidence that our tests are actually testing the queuing code path. For
> instance, because most tests just client.submit a couple tasks, a ton of
> tests were actually not using the queuing code path even with worker-saturation:
> 1.0. (This change is how I found #7223
> <https://github.com/dask/distributed/issues/7223>).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/distributed/pull/7221#issuecomment-1297605957>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACKZTB55UWUVSJ2SX23KUTWGAQITANCNFSM6AAAAAARRKIJHU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gjoseph92 commented:
- https://github.com/dask/distributed/issues/7208 everywhere
- https://github.com/dask/distributed/issues/6820
- `distributed/tests/test_cluster_dump.py::test_cluster_dump_to_yamls` but I've seen this before, fixed in https://github.com/dask/distributed/pull/6015 5bbaaa49b78e71b5c9036060982cac9ed8f2d0c6

`distributed/tests/test_scheduler.py::test_balance_many_workers - assert {0, 1, 2} == {0, 1}` could be real, looking into it: https://github.com/dask/distributed/actions/runs/3364213017/jobs/5578308124#step:18:1315

This makes me wonder if the result is very meaningful though:
````
2022-10-31 20:08:16,899 - distributed.utils_perf - WARNING - full garbage collections took 24% CPU time recently (threshold: 10%)
2022-10-31 20:08:17,803 - distributed.utils_perf - WARNING - full garbage collections took 24% CPU time recently (threshold: 10%)
2022-10-31 20:08:40,283 - distributed.core - INFO - Event loop was unresponsive in Scheduler for 24.96s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
````

gjoseph92 commented:
All green besides:
* https://github.com/dask/distributed/issues/7208
* `distributed/dashboard/tests/test_scheduler_bokeh.py::test_shuffling`

crusaderky commented on distributed/scheduler.py:
```diff
@@ -2119,11 +2119,13 @@ def decide_worker_rootish_queuing_disabled(
         )
-        tg.last_worker_tasks_left -= 1
+        if tg.last_worker_tasks_left > 0:
+            tg.last_worker_tasks_left -= 1
```
To my understanding, this covers the use case where
`(len(tg) / self.total_nthreads) * ws.nthreads` < 1.
Since a task is rootish iff `len(tg) > total_nthreads * 2`, this should be mathematically impossible.

gjoseph92 replied:
> Since a task is rootish iff `len(tg) > total_nthreads * 2`

The main change in this PR is that this is no longer the case. Since we now call any task with 0 dependents root-ish—regardless of the size of its TaskGroup—it's possible for `len(tg)` to be as small as 1.

crusaderky replied:
Ah, I see. I refactored the code to add clarity.

crusaderky commented on distributed/tests/test_priorities.py:
```diff
@@ -51,48 +57,40 @@ async def block_worker(
-            pytest.param(True, id="queue on scheduler"),
-        ],
-    )(
+    return pytest.mark.parametrize("pause", [pytest.param(False), pytest.param(True)])(
```
```suggestion
    return pytest.mark.parametrize("pause", [False, True])(
```

crusaderky reviewed (changes requested):
I think there's a test for an impossible condition that should be replaced with an assertion (read above). Everything else looks good.

gjoseph92 commented:
> there's a test for an impossible condition

@crusaderky see https://github.com/dask/distributed/pull/7221#discussion_r1012407756

gjoseph92 commented:
Again only failure is
* https://github.com/dask/distributed/issues/7208

gjoseph92 commented:
Follow-ups if this is merged:
* https://github.com/dask/distributed/pull/7245
* https://github.com/dask/distributed/pull/7222

gjoseph92 commented:
Another reason to merge this:

The current code path (when the taskgroup is smaller than the cluster) can have bad behavior for `worker-saturation > 1.0`. Because of the round-up, when using a value like 1.1, workers will be in the `idle` set even when all their threads are in use. In a non-homogeneous cluster, this can lead to picking a worker that's completely full even when there are workers with open threads. That's the original thing that happened in https://github.com/dask/distributed/issues/7197.

fjetter commented on distributed/tests/test_client_executor.py:
```diff
@@ -213,18 +213,13 @@ def test_unsupported_arguments(client, s, a, b):
 
-    with client.get_executor(retries=5, pure=False) as e:
+    with client.get_executor(retries=6, pure=False) as e:
         future = e.submit(varying(args))
         assert future.result() == 42
 
-    with client.get_executor(retries=4) as e:
+    with client.get_executor(retries=1) as e:
         future = e.submit(varying(args))
-        result = future.result()
-        assert result == 42
-
-    with client.get_executor(retries=2) as e:
-        future = e.submit(varying(args))
-        with pytest.raises(ZeroDivisionError, match="two"):
```
Any reason to change these retries?

gjoseph92 replied:
This test is extremely sensitive to how workers are selected on an idle cluster. It uses this stateful `varying` utility, which changes its behavior depending on how many times it's been called on that worker.

It had to be changed when the idle-round-robin behavior was added: https://github.com/dask/distributed/pull/4638/files#diff-59af67191283f0c64a3be8ce1f344f49b9d025f8264b77fba5c8250865bde433

So I've had to change it again since it's being removed.

fjetter commented on distributed/tests/test_priorities.py:
```diff
@@ -51,48 +57,40 @@ async def block_worker(
-    return pytest.mark.parametrize(
-        "pause",
-        [
-            pytest.param(False, id="queue on worker"),
-            pytest.param(True, id="queue on scheduler"),
-        ],
```
Why remove the ids?

gjoseph92 replied:
They weren't accurate anymore. I could call then `pause` and `clog` or something.

fjetter commented on distributed/tests/test_scheduler.py:
```diff
@@ -2849,6 +2860,10 @@ async def test_get_worker_monitor_info(s, a, b):
         assert res[w.address]["last_time"] is not None
 
 
+@pytest.mark.skipif(
```
I think we need to preserve this kind of behavior. I don't care if it's actually round robin or not but if I submit three tasks and there are two workers, both workers should have work.

What's the plan to fix this?

gjoseph92 replied:
> if I submit three tasks and there are two workers, both workers should have work

That's not what this is testing. What you're describing above already will happen. This PR in fact _improves_ that case. See https://github.com/dask/distributed/issues/7197: if you set worker-saturation to 1.1 right now, `test_wait_first_completed` will fail because 2 tasks get assigned to the 1-threaded worker, and one task is assigned to the 2-threaded worker, because they're using the old code path.

This test is testing that if you submit a task to a completely empty cluster, wait for it to complete, release it, then submit another task, that you'll get different workers each time. That's a different and much more niche case.

Approaches for this:
* https://github.com/dask/distributed/pull/7222
* https://github.com/dask/distributed/pull/7248

fjetter replied:
Please delete the test

gjoseph92 commented:
> You are already discussing four different edge cases (and I'm sure there are more). Can you please write them as a test? This way we know for a fact that the code behaves this way instead of relying on a lengthy argument.

I'm not following what you'd want to assert about the cases mentioned in here https://github.com/dask/distributed/pull/7221#issuecomment-1297605957 (I'm assuming those are the four cases you're talking about).

There are already a number of tests that cover aspects of this behavior, for instance:

https://github.com/dask/distributed/blob/02b9430f7bb6eae370394eaac6b4e8f574746c94/distributed/tests/test_client.py#L3255-L3260

https://github.com/dask/distributed/blob/02b9430f7bb6eae370394eaac6b4e8f574746c94/distributed/tests/test_client.py#L3272-L3284

https://github.com/dask/distributed/blob/02b9430f7bb6eae370394eaac6b4e8f574746c94/distributed/tests/test_scheduler.py#L1543-L1565

I could see adding tests for:
1. `client.submit` after the cluster is already saturated doesn't get to cut in line before tasks that were submitted earlier (would fail today)
2. `client.submit` in a for loop with follow-up tasks doesn't have overproduction with queuing on. This would just be [`test_graph_execution_width`](https://github.com/dask/distributed/blob/4c684a276a3c01b101c66fd11d223cce1a1356a1/distributed/tests/test_scheduler.py#L329-L369) using futures instead of delayed.

fjetter commented:
I'm cool with merging iff
1. Find out where we hit the "dead code". If it's actually dead code we need to remove it. If it is not we should have a test that shows the difference in behavior
2. Have a guestimate on how bad performance is
3. CI is green-ish

gjoseph92 commented:
I removed the dead code in https://github.com/dask/distributed/pull/7221/commits/c9018239266366ef980803c55419b91d4485c3a8 and added an assertion. Turns out there was a place it was hit. I'm addressing that in https://github.com/dask/distributed/pull/7259, which will need to be merged first.

gjoseph92 commented:
> Have a guestimate on how bad performance is

Ran a benchmark on a larger cluster (not with this PR, but still relevant): https://github.com/dask/distributed/issues/7246#issuecomment-1302889869

tl;dr performance impact is hardly measurable.

gjoseph92 commented:
I'm wondering if a better way to solve the consistency issue is to just cache root-ish-ness so we don't have to worry about it changing: https://github.com/dask/distributed/pull/7262

crusaderky commented on distributed/scheduler.py:
```diff
@@ -2014,6 +2014,8 @@ def transition_no_worker_processing(self, key, stimulus_id):
 
             if ws := self.decide_worker_non_rootish(ts):
+                # ^ NOTE: `ts` may actually be root-ish now, but it wasn't when it went
+                # into `no-worker`. `TaskGroup` or cluster size could have changed.
```
As mentioned in https://github.com/dask/distributed/pull/7259/files#r1015290381 this should be impossible.
